### PR TITLE
Added new command 'summary'

### LINKED ISSUE #1095: Add default `frictionless` CLI command
# Overview

Currently, we have these commands:
- `frictionless describe`
- `frictionless extract`
- `frictionless validate`
- [coming] `frictionless analyze`
- etc

We'd like to have (note there is no action):

```
$ frictionless table.csv

# Describe

[resource/schema as a table]

# Extract

[first X lines of the data as a table]

# Validate

## Summary

[validation summary table]

## Errors 

[validation errors table]
```

### PR DESCRIPTION
**Pull request by shashigharti**
* Added new command named 'summary'.
* Added to_summary function to ReportTask.
* Added tests for both 'summary' command and 'to_summary' function.
* Reorganized the commonly used function to helpers.

- fixes #1095

- Shall I also update the documentation here 
- https://framework.frictionlessdata.io/docs/guides/validation-guide/#validation-errors to add  "to_summary" info.

### CODE REVIEW

**Comment by shashigharti** [comment]
The output looks like this:
`summary data/countries.csv`
output

**For to_summary**
```
for task in report.tasks:
    summary = task.to_summary()
    print("# Describe \n")
    print(summary['describe'])
    print("\n# Extract \n")
    print(summary['extract'])
    print("\n# Validate \n")
    print(task.valid)
    print("\n# Summary \n")
    print(summary['validate']['summary'])
    print("\n# Errors \n")
    print(summary['validate']['errors'])
```
```
# Describe 

+-------------+---------+------------+
| name        | type    | required   |
+=============+=========+============+
| id          | integer |            |
+-------------+---------+------------+
| neighbor_id | string  |            |
+-------------+---------+------------+
| name        | string  |            |
+-------------+---------+------------+
| population  | string  |            |
+-------------+---------+------------+

# Extract 

+----+-------------+-----------+------------+
| id | neighbor_id | name      | population |
+====+=============+===========+============+
|  1 | 'Ireland'   | 'Britain' | '67'       |
+----+-------------+-----------+------------+
|  2 | '3'         | 'France'  | 'n/a'      |
+----+-------------+-----------+------------+
|  3 | '22'        | 'Germany' | '83'       |
+----+-------------+-----------+------------+
|  4 | None        | 'Italy'   | '60'       |
+----+-------------+-----------+------------+
|  5 | None        | None      | None       |
+----+-------------+-----------+------------+


# Validate 

False

# Summary 

+-----------------------------+-------------------------------------------------------------+
| Description                 | Size/Name/Count                                             |
+=============================+=============================================================+
| File name                   | /home/shashi/python/okfn/frictionless-py/data/countries.csv |
+-----------------------------+-------------------------------------------------------------+
| File size (bytes)           | 143                                                         |
+-----------------------------+-------------------------------------------------------------+
| Total Time Taken (sec)      | 0.007                                                       |
+-----------------------------+-------------------------------------------------------------+
| Total Errors                | 4                                                           |
+-----------------------------+-------------------------------------------------------------+
| Extra Cell (extra-cell)     | 1                                                           |
+-----------------------------+-------------------------------------------------------------+
| Missing Cell (missing-cell) | 3                                                           |
+-----------------------------+-------------------------------------------------------------+

# Errors 

+-------+---------+------------+----------------------------------------------------+
|   row |   field | code       | message                                            |
+=======+=========+============+====================================================+
|     4 |       5 | extra-cell | Row at position "4" has an extra value in field at |
|       |         |            | position "5"                                       |
+-------+---------+------------+----------------------------------------------------+
|     7 |       2 | missing-   | Row at position "7" has a missing cell in field    |
|       |         | cell       | "neighbor_id" at position "2"                      |
+-------+---------+------------+----------------------------------------------------+
|     7 |       3 | missing-   | Row at position "7" has a missing cell in field    |
|       |         | cell       | "name" at position "3"                             |
+-------+---------+------------+----------------------------------------------------+
|     7 |       4 | missing-   | Row at position "7" has a missing cell in field    |
|       |         | cell       | "population" at position "4"                       |
+-------+---------+------------+----------------------------------------------------+
```

File: frictionless/report/report.py [hunk 1]
```diff
@@ -331,6 +333,82 @@ def flatten(self, spec=["rowPosition", "fieldPosition", "code"]):
             result.append([context.get(prop) for prop in spec])
         return result
 
+    # Summary
+
+    def to_summary(self) -> dict:
```
Comment (roll) [inline frictionless/report/report.py]: @shashigharti 
It looks really nice in-general!

I would think more closely about separation of concerns - https://en.wikipedia.org/wiki/Separation_of_concerns

It's great to have `report.to_summary` (or `report_task.to_summary`) but in my opinion it should be only related to the validation details (and includes helpers.validation_summary code)

For describe we can have `schema.to_summary`.

And for extract something like `resource.to_summary`

Then composing everything in the `program.summary`:
- schema.to_summary() # describe
- resource.to_summary() # extract
- report.to_summary() # validate

**Review by roll — CHANGES_REQUESTED** [review]

**Comment by shashigharti** [comment]
@roll thank you again!. I have revised the code and put in in the relevant module as you have suggested. And also  changed/added tests. 

I did not add to_summary function to resource because there is already a function to_view() and I used it to display the data.

I am not sure if  tests > resource > validate >test_general is the right place to add the test for schema.to_summay?

File: frictionless/helpers.py [hunk 2]
```diff
@@ -634,10 +634,62 @@ def dicts_to_markdown_table(dicts: List[dict], **kwargs) -> str:
...
 def format_bytes(size: int) -> str:
     """Format bytes to larger units"""
     units = ["bytes", "KB", "MB", "GB", "TB"]
     index = math.floor(math.log2(size) / 10)
     if index > len(units):
         index = len(units) - 1
     return units[index]
+
+
+def validation_summary(
```
Comment (roll) [inline frictionless/helpers.py]: Can we also move it to somewhere in `report.py` (a method or a helper)?

**Review by roll — CHANGES_REQUESTED** [review]
Great!

I've added a minor comment.

Also, answering your test question -- if it's a report method tests should be in `tests/report`, schema method test should be in `tests/schema` etc

File: frictionless/helpers.py (hunk 2 above)
Comment (shashigharti) [reply frictionless/helpers.py]: @roll thanks. It is being used by validate command also. We might have to change that command also in that case.

Comment (roll) [reply frictionless/helpers.py]: Yes I was thinking if we can have this method as a part of `report` or `reportTask`

Comment (shashigharti) [reply frictionless/helpers.py]: Thank you! I have made it part of report class.

**Comment by shashigharti** [comment]
Thank you @roll ,  I have made the changes and it is ready for the review. I also placed _wrap_text_to_colwidths function in report.py file and replaced the validate report table display in validate.py command to report.to_summary().

https://github.com/frictionlessdata/frictionless-py/pull/1127/commits/e936c42a7b3a891569a762db8d5c1fa897bf221c#diff-ca29842cf90a4bca29af564ebbb9936fce52dfb7d085a7dfe10d58124c991c65R218

what do you think of it?

File: frictionless/report/report.py [hunk 3]
```diff
@@ -166,6 +170,119 @@ def wrapper(*args, **kwargs):
...
+                        error_content,
+                        headers=["row", "field", "code", "message"],
+                        tablefmt="grid",
+                    )
+                )
+                validation_content += "\n\n"
+
+        return validation_content
+
+    def validation_summary(
```
Comment (roll) [inline frictionless/report/report.py]: @shashigharti 
I think the last step will be moving this method to be `reportTask.to_summary` so you don't need to pass any arguments? The description says "Generate summary for validation task" which is a clear indicator that "self" in this method should be a `reportTask` object I guess

**Review by roll — CHANGES_REQUESTED** [review]

File: frictionless/report/report.py (hunk 3 above)
Comment (shashigharti) [reply frictionless/report/report.py]: Yes, most of metrics for validation summary can be calculated using information from reportTask so I have rearranged the code to reportTask.to_summary as you suggested.

**Comment by shashigharti** [comment]
Thank you @roll . I have made the changes. I have also removed the test (that I had added) that were not very related to the changes.

**Review by roll — APPROVED** [review]
Great!

**Pull request merged**